### PR TITLE
Work on ctype functions, fix some string functions

### PR DESCRIPTION
**Pull request by carbotaniuman**
I'm working on getting [the bat binary](https://github.com/sharkdp/bat) to run on Rustix.

The main changes in this PR are:
- An implementation of all <ctype.h> functions not dependent on locale
- Glibc ABI shenanigans for ctype stuff
- strcasecmp and strncasecmp
- the 4 ip endian switching functions
- strerror + some libc delegation fixes for string functions

### CODE REVIEW

File: c-scape/src/ctype/lsb_abi/ctype_b_loc.rs [hunk 1]
```diff
@@ -0,0 +1,399 @@
+use super::SyncWrapper;
+
+pub const fn bsw(v: u16) -> u16 {
+    u16::from_be(v)
+}
```
Comment (sunfishcode) [inline c-scape/src/ctype/lsb_abi/ctype_b_loc.rs:5]: If rustfmt is going to put each element on its own line, would it make sense to just inline this?

**Review by sunfishcode — COMMENTED** [review]
The `bat` binary is a great idea :-). Thanks for working on this!

**Comment by carbotaniuman** [comment]
I can't seem to figure out how to get the alignment on x86 to be correct.

**Comment by sunfishcode** [comment]
The alignment appears to be 1 on i686-unknown-linug-gnu, and is overridden to be higher on specific configurations: https://github.com/rust-lang/libc/blob/master/src/unix/linux_like/linux/align.rs#L31

**Comment by carbotaniuman** [comment]
TIL `rustix::io::OwnedFd` and `rustix::fd::OwnedFd` are not the same...

**Comment by sunfishcode** [comment]
Yeah. and it's just `OwnedFd` that has a wrapper in `rustix::io` btw, because the issue is the `close` call that it does. For the others we just use `rustix::fd` directly.

File: c-scape/src/sync_ptr.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/// A const pointer to `T` that implements [Sync].
```
Comment (sunfishcode) [inline c-scape/src/sync_ptr.rs:1]: Pplease put the `Sync` in code quotes.

File: c-scape/src/sync_ptr.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    /// Creates a `SyncConstPointer` from a raw pointer.
+    ///
+    /// Behavior is undefined if `ptr` is actually not
+    /// safe to share across threads.
+    pub const unsafe fn new(ptr: *const T) -> Self {
+        Self(ptr)
+    }
+}
+
+/// A mut pointer to `T` that implements [Sync].
```
Comment (sunfishcode) [inline c-scape/src/sync_ptr.rs:17]: Here too.

File: c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    'h' as i32, 'i' as i32, 'j' as i32, 'k' as i32, 'l' as i32, 'm' as i32, 'n' as i32, 'o' as i32,
+    'p' as i32, 'q' as i32, 'r' as i32, 's' as i32, 't' as i32, 'u' as i32, 'v' as i32, 'w' as i32,
+    'x' as i32, 'y' as i32, 'z' as i32, 123, 124, 125, 126, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+];
+
+static PTR: SyncUnsafeCell<SyncConstPtr<i32>> =
```
Comment (sunfishcode) [inline c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs:27]: How does using `SyncUnsafeCell` here differ from using a plain mutable static?

**Review by sunfishcode — COMMENTED** [review]
Looks good! Just a few more minor nits and questions.

File: c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs (hunk 4 above)
Comment (carbotaniuman) [reply c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs:27]: I think `SyncUnsafeCell` is more ergonomic and less footgunny than mutable statics, such as even getting a pointer to a mutable static with `addr_of` needing unsafe.

File: c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    SyncUnsafeCell::new(unsafe { SyncConstPtr::new(addr_of!(TABLE[128])) });
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs:28]: Oh, wow. Cool.

File: c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs (hunk 4 above)
Comment (sunfishcode) [reply c-scape/src/ctype/lsb_abi/ctype_tolower_loc.rs:27]: Sounds good. I also like how it makes it clear that these statics really are mutated in unsafe ways.

**Pull request merged**